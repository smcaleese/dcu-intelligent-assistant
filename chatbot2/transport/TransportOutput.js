import React, { useState } from 'react';
import { StyleSheet, View, Text, ScrollView} from 'react-native';
import BusTile from '../components/BusTile';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import CustomButton from '../components/CustomButton';
import { getBusByStop, getBusByNumber } from '../utils/GetBus';

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#f0f0f0',
        alignItems: 'center'
    },
    titleContainer: {
        backgroundColor: "#4269E1",
        width: "95%",
        borderTopLeftRadius: 10,
        borderTopRightRadius: 10,
        padding: 5,
        marginTop: 50
    },
    titleText: {
        textAlign: 'center',
        fontSize: 24,
        color: "white"
    },
    bg: {
        marginTop: 0,
        backgroundColor: 'white',
        borderBottomLeftRadius: 10,
        borderBottomRightRadius: 10,
        width: '95%',
        justifyContent: 'center',
        alignItems: 'center',
    },
    scrollView: {
        width: "100%",
        height: "65%"
    }, 
    stopTitleContainer: {
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'center',
        borderBottomColor: '#d3d3d3',
        borderBottomWidth: 1.5,
        margin: 10,
        marginBottom: 16
    },
    stopTitle: {
        color: "#535353",
        padding: 2,
        fontSize: 18,
        textAlign: 'center',
        width: '80%'
    },
    stopicon: {
        textAlign: 'center',
        width: "20%"
    },
    button: {
        width: '60%',
        margin: 10,
        borderRadius: 5,
        padding: 15
    },
    buttonText: {
        color: 'white',
        fontSize: 16,
        textAlign: 'center'
    }
})

export default function TransportOutput(props) {
    //Sort the Array of inputs but time diffrence (Which one is arriving next)
    const [timeState, setTimeState] = useState(props.route.params.times.sort((a,b) => a.time_diff > b.time_diff ? 1 : -1))

    // Reduce function from https://stackoverflow.com/questions/40774697/how-to-group-an-array-of-objects-by-key
    //Groups all the bus times in the array by stop
    const groupedByStop = timeState.reduce((r, time) => {
        r[time.stop_name] = r[time.stop_name] || [];
        r[time.stop_name].push(time);
        return r;
    }, Object.create(null))

    //Allows a user to update the results
    const handleRefresh = () => {
        if (props.route.params.type === "bus"){
            getBusByNumber(props.route.params.input)
            .then((res) => {
                setTimeState(res.sort((a,b) => a.time_diff > b.time_diff ? 1 : -1))
            })
            .catch((err) => console.log(err))
        }
        else{
            getBusByStop(props.route.params.input)
            .then((res) => {
                setTimeState(res.sort((a,b) => a.time_diff > b.time_diff ? 1 : -1))
            })
            .catch((err) => console.log(err))
        }
    }

    console.log("groupedByStop:", groupedByStop)

    return (
        <View style={styles.container}>
            <View style={styles.titleContainer}>
                <Text style={styles.titleText}>Bus Times</Text>
            </View>
            <View style={styles.bg}>
                {timeState.length ?
                <>
                    <ScrollView style={styles.scrollView}>
                        <View>
                        {Object.keys(groupedByStop).map((stop, index) => {
                                return(
                                    <View key={index}>
                                        <View style={styles.stopTitleContainer}>
                                            <Text style={styles.stopTitle}>{stop}</Text>
                                        </View>
                                        <View>
                                        {groupedByStop[stop].map((time, index) => {
                                            return(
                                                <BusTile key={index} stop={time} />
                                            );
                                            })
                                        }
                                        </View>
                                    </View>
                                );
                            })}
                        </View>
                    </ScrollView>
                    <View style={{margin: 10}}>
                        <CustomButton title="Refresh" color="#4269E1" onPress={handleRefresh} />
                    </View>
                    </>
                :
                    <Text style={styles.stopTitle}>Sorry no buses arriving in the next hour. Please try again later.</Text>
                }
            </View>
        </View>
    )
}