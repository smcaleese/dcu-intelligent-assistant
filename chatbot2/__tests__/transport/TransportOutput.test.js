import React from 'react';
import {create} from 'react-test-renderer';
import TransportOutput from '../../transport/TransportOutput';

// "2021-03-10T15:30:30.000Z"

const params = {
    params: {
        times: [
            {   
                arrival_time: "2021-03-10T15:30:30.000Z",
                trip_headsign: "Test-Dest",
                bus_number: "123",
                time_diff: "10",
                stop_name: "Test"
            },
            {   
                arrival_time: "2021-03-10T15:40:30.000Z",
                trip_headsign: "Test-Dest",
                bus_number: "123",
                time_diff: "20",
                stop_name: "Test"
            },
        ]
    }
}

jest.useFakeTimers();

const screen = create(<TransportOutput route={params} />)

test('TransportOutput Snapshot Test', () => {
    expect(screen).toMatchSnapshot();
})