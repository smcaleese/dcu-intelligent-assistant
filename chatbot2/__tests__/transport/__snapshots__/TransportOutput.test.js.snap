// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransportOutput Snapshot Test 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f0f0f0",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#4269E1",
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "marginTop": 50,
        "padding": 5,
        "width": "95%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 24,
          "textAlign": "center",
        }
      }
    >
      Bus Times
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderBottomLeftRadius": 10,
        "borderBottomRightRadius": 10,
        "justifyContent": "center",
        "marginTop": 0,
        "width": "95%",
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "height": "65%",
          "width": "100%",
        }
      }
    >
      <View>
        <View>
          <View>
            <View
              style={
                Object {
                  "borderBottomColor": "#d3d3d3",
                  "borderBottomWidth": 1.5,
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "margin": 10,
                  "marginBottom": 16,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#535353",
                    "fontSize": 18,
                    "padding": 2,
                    "textAlign": "center",
                    "width": "80%",
                  }
                }
              >
                Test
              </Text>
            </View>
            <View>
              <View>
                <View
                  style={
                    Object {
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "margin": 5,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "backgroundColor": "#4269E1",
                        "borderBottomLeftRadius": 5,
                        "borderTopLeftRadius": 5,
                        "height": 70,
                        "justifyContent": "center",
                        "padding": 5,
                        "width": 70,
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "black",
                            "fontWeight": "bold",
                            "textAlign": "center",
                            "textAlignVertical": "center",
                          },
                          Object {
                            "color": "white",
                          },
                        ]
                      }
                    >
                      123
                    </Text>
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": "#FFF",
                            "fontSize": 24,
                          },
                          Object {
                            "textAlign": "center",
                            "width": "100%",
                          },
                          Object {
                            "fontFamily": "Material Design Icons",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                    >
                      󱀒
                    </Text>
                  </View>
                  <View
                    style={
                      Object {
                        "backgroundColor": "white",
                        "borderColor": "#d3d3d3",
                        "borderLeftWidth": 0,
                        "borderRightWidth": 0,
                        "borderWidth": 2,
                        "height": 70,
                        "justifyContent": "center",
                        "margin": 0,
                        "overflow": "hidden",
                        "padding": 5,
                        "width": "45%",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "black",
                            "fontWeight": "bold",
                            "textAlign": "center",
                            "textAlignVertical": "center",
                          },
                          Object {
                            "fontSize": 12,
                          },
                        ]
                      }
                    >
                      Test-Dest
                    </Text>
                    <Text
                      style={
                        Object {
                          "fontSize": 12,
                          "textAlign": "center",
                        }
                      }
                    >
                      15:30
                    </Text>
                  </View>
                  <View
                    style={
                      Object {
                        "backgroundColor": "#4269E1",
                        "borderBottomRightRadius": 5,
                        "borderTopRightRadius": 5,
                        "flexDirection": "column",
                        "height": 70,
                        "justifyContent": "center",
                        "padding": 5,
                        "width": 70,
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "color": "white",
                          "fontWeight": "bold",
                          "textAlign": "center",
                        }
                      }
                    >
                      10
                    </Text>
                    <Text
                      style={
                        Object {
                          "color": "white",
                          "fontWeight": "bold",
                          "textAlign": "center",
                        }
                      }
                    >
                      mins
                    </Text>
                  </View>
                </View>
              </View>
              <View>
                <View
                  style={
                    Object {
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "margin": 5,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "backgroundColor": "#4269E1",
                        "borderBottomLeftRadius": 5,
                        "borderTopLeftRadius": 5,
                        "height": 70,
                        "justifyContent": "center",
                        "padding": 5,
                        "width": 70,
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "black",
                            "fontWeight": "bold",
                            "textAlign": "center",
                            "textAlignVertical": "center",
                          },
                          Object {
                            "color": "white",
                          },
                        ]
                      }
                    >
                      123
                    </Text>
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": "#FFF",
                            "fontSize": 24,
                          },
                          Object {
                            "textAlign": "center",
                            "width": "100%",
                          },
                          Object {
                            "fontFamily": "Material Design Icons",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                    >
                      󱀒
                    </Text>
                  </View>
                  <View
                    style={
                      Object {
                        "backgroundColor": "white",
                        "borderColor": "#d3d3d3",
                        "borderLeftWidth": 0,
                        "borderRightWidth": 0,
                        "borderWidth": 2,
                        "height": 70,
                        "justifyContent": "center",
                        "margin": 0,
                        "overflow": "hidden",
                        "padding": 5,
                        "width": "45%",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "black",
                            "fontWeight": "bold",
                            "textAlign": "center",
                            "textAlignVertical": "center",
                          },
                          Object {
                            "fontSize": 12,
                          },
                        ]
                      }
                    >
                      Test-Dest
                    </Text>
                    <Text
                      style={
                        Object {
                          "fontSize": 12,
                          "textAlign": "center",
                        }
                      }
                    >
                      15:40
                    </Text>
                  </View>
                  <View
                    style={
                      Object {
                        "backgroundColor": "#4269E1",
                        "borderBottomRightRadius": 5,
                        "borderTopRightRadius": 5,
                        "flexDirection": "column",
                        "height": 70,
                        "justifyContent": "center",
                        "padding": 5,
                        "width": 70,
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "color": "white",
                          "fontWeight": "bold",
                          "textAlign": "center",
                        }
                      }
                    >
                      20
                    </Text>
                    <Text
                      style={
                        Object {
                          "color": "white",
                          "fontWeight": "bold",
                          "textAlign": "center",
                        }
                      }
                    >
                      mins
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "margin": 10,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#4269E1",
            "borderRadius": 5,
            "margin": 10,
            "opacity": 1,
            "padding": 15,
            "width": 200,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          Refresh
        </Text>
      </View>
    </View>
  </View>
</View>
`;
